src/main/java/org/wildfly/security/util/HttpClient.java

```diff
@@ -0,0 +1,284 @@
+
+        /**
+         * <p>Executes this request and notifies the specified <code>responseHandler</code> about the response sent by the server.
+         *
+         * <p>The <code>responseHandler</code> will also be notified about errors returned from the server. In this case,
+         * implementations may expect a {@link Throwable} with additional information about the error as well a {@link InputStream}
+         * with additional information returned by the server.
+         *
+         * @param responseHandler the response handler
+         * @return the response handler result
+         * @throws RuntimeException in case any unexpected error occurs when executing the request
+         */
+        public <R> R execute(BiFunction<InputStream, Throwable, R> responseHandler) throws RuntimeException {
+            checkNotNullParam("responseHandler", responseHandler);
+            HttpURLConnection connection = null;
+
+            try {
+                connection = openConnection();
+                connection.setDoOutput(true);
+                connection.setRequestMethod(method);
+
+                for (HttpHeader header : headers) {
+                    connection.setRequestProperty(header.getName(), header.getValue());
+                }
+
+                if (METHOD_POST.equals(method)) {
+                    byte[] paramBytes = buildParameters(params);
+
+                    try (OutputStream outputStream = connection.getOutputStream()) {
+                        outputStream.write(paramBytes);
+                    }
+                }
+
+                try (InputStream inputStream = new BufferedInputStream(connection.getInputStream())) {
+                    return responseHandler.apply(inputStream, null);
+                }
+            } catch (IOException ioe) {
+                InputStream errorStream = null;
+
+                if (connection != null && connection.getErrorStream() != null) {
+                    errorStream = connection.getErrorStream();
+
+                    try (BufferedReader reader = new BufferedReader(new InputStreamReader(errorStream))) {
+                        StringBuffer response = reader.lines().reduce(new StringBuffer(), StringBuffer::append, (buffer1, buffer2) -> buffer1);
+                        log.errorf(ioe, "Unexpected response from server [%s]. Response: [%s]", url, response);
+                    } catch (IOException e) {
+                        return responseHandler.apply(errorStream, e);
+                    }
+                }
+
+                return responseHandler.apply(errorStream, ioe);
+            } catch (Exception e) {
+                throw log.httpClientUnexpectedResponseFromServer(e);
+            }
+        }
+
+        private HttpURLConnection openConnection() throws IOException {
+            try {
+                log.debugf("Opening connection to server [%s]", url);
+                HttpURLConnection connection = (HttpURLConnection) url.openConnection();
```

Comment:
I think this is the point we may want to be sure we don't automatically follow redirects from http to https.